designs/disruption-controls.md

```diff
@@ -0,0 +1,118 @@
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    budgets:
+    # On Weekdays during business hours, don't do any deprovisioning.
+    - crontab: "0 9 * * mon-fri"
+      duration: 8h
+      maxUnavailable: 0
+    # Every other time, only allow 10 nodes to be deprovisioned simultaneously
+    - maxUnavailable: 10
+```
+
+## Code Definition
+
+```go
+type Disruption struct {
+    {...}
+    // Budgets is a list of Budgets.
+    // If there are multiple active budgets, the most restrictive budget's maxUnavailable is respected.
+    Budgets []Budget `json:"budgets,omitempty" hash:"ignore"`
+    // ConsolidateAfter is a nillable duration, parsed as a metav1.Duration.
+    // Users can use "Never" to disable Consolidation.
+    ConsolidateAfter *NillableDuration `json:"consolidateAfter" hash:"ignore"`
+    // ConsolidationPolicy determines how Karpenter will consider nodes
+    // as candidates for Consolidation.
+    // WhenEmpty uses the same behavior as v1alpha5 TTLSecondsAfterEmpty
+    // WhenUnderutilized uses the same behavior as v1alpha5.ConsolidationEnabled: true
+    ConsolidationPolicy string `json:"consolidationPolicy" hash:"ignore"`
+}
+// Budget specifies periods of times where Karpenter will restrict the
+// number of Node Claims that can be terminated at a time.
+// Unless specified, a budget is always active.
+type Budget struct {
+    // MaxUnavailable dictates how many Node Claims owned by this NodePool
+    // can be terminating at once.
+    // This only respects and considers nodes with a DeletionTimestamp set.
+    MaxUnavailable intstr.IntOrString `json:"maxUnavailable" hash:"ignore"`
+    // Crontab specifies when a budget begins being active.
+    // Crontab uses the same syntax as a Cronjob:
+    // "Minute Hour DayOfMonth Month DayOfWeek"
+    // This is required if Duration is set.
+    Crontab *string `json:"crontab,omitempty" hash:"ignore"`
+    // Duration determines how long a Budget is active since each Crontab hit.
+    // This is required if Crontab is set.
+    Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+}
+```
+
+## Validation/Defaults
+
+For each `Budget`, users must set a non-negative `MaxUnavailable`. Users can disable scale down for a NodePool by setting this to `0`. Users must either omit both `Crontab` and `Duration` or set both of them, since `Crontab` and `Duration` are inherently linked. Omitting these two fields will be equivalent to an always active `Budget`. Users cannot define a seconds value in `Duration`, since the smallest denomination of time in upstream Crontabs are minutes.
```

Comment:
The document does not define how either the numerator or denominator of the budget are calculated.

Reply:
Ah good point! You're talking about the case of using a percent with `MaxUnavailable`, right? 

Let's say if a user had 10 nodes, but used a `maxUnavailable` of 52%, that would mean Karpenter could disrupt at most 5.2 nodes at a time. Since `.2` nodes doesn't make sense, `6 nodes` would be violating that budget, but `5` wouldn't be. So it would effectively round down the value. 

Is this what you're talking about?

Reply:
I'm saying you're not defining what is counted to arrive at the number `10` and the number that can't be more than `5`.

Reply:
Karpenter will only be aware of nodes that have the node pool name label on and are created by Karpenter. So 10 in this case would be the number of nodes already created by Karpenter for this node pool. The number that we'll check in reference is how many of those nodes have a non-nil deletion timestamp, meaning they're actively being deleted.

Reply:
That information should be put in the document.

I would suggest cordoning as a better signal than the deletion timestamp.

Reply:
By "nodes" do you mean Kubernetes `Node` objects or does it include instances that have been created by Karpenter but which have not yet joined the cluster?

Reply:
If there are 6 nodes of 10 without a deletion timestamp and 3 of those 6 are in NotReady state, that allows an additional potentially-Ready node to be drift-remediated?

Reply:
Will do. The "cordoned" signal is a little tougher to reason about. It's possible that a node could be cordoned by another controller within the cluster, that we shouldn't consider. Right now, Karpenter uses the upstream `kubernetes.io/unschedulable` taint, which many other controllers and pods are aware of. Yet, just because a node is cordoned doesn't mean it's being disrupted. We're exploring changing this to another `disrupting` taint, which I think could be a better way for Karpenter to be aware of this. 

If we migrate to a new Karpenter-specific taint, you're right. I think that would probably the better way to drive it. 


> By "nodes" do you mean Kubernetes Node resources or does it include instances that have been created by Karpenter but which have not yet joined the cluster?

We will only be considering nodes in the cluster that have an associated NodeClaim, and vice versa. We only continue with disruptions at the moment if nodes/nodeClaims are initialized and have been registered with Karpenter.

> If there are 6 nodes of 10 without a deletion timestamp and 3 of those 6 are in NotReady state, that allows an additional potentially-Ready node to be drift-remediated?

Yeah, we should be including ways to handle nodes that are NotReady for an extended period of time, but I think regardless of if a node is Ready or not, if it's drifted or expired, I don't see why we shouldn't try to replace it if the budget allows it. Are you thinking that the 3 NotReady nodes in this case should also contribute to the budget?

Reply:
I'm thinking that disruption should not progress until sufficient new nodes are at least in the Ready state; nodes in NotReady state probably should not count as satisfying the budget.

There is a gap between the "Ready" state and able to take workloads, but PodDisruptionBudgets should suffice to cover that. I see the node disruption budget as covering churn in supporting systems, such as the Kubernetes API, CNIs, monitoring systems, etc.

You might want to consider the gap between a new node becoming Ready and the end of elevated activity due to workloads piling onto it.

Reply:
If it's `NodeClaims` that are being counted for the `10`, that should be what is documented. Or if it's the intersection of `Nodes` and `NodeClaims`, likewise.

Reply:
I've added in that it's tracked based off nodeClaims and that it only refers to nodes with the `karpenter.sh/disruption` taint.